# Feature selection and feature engineering

This isn't based on a news article, exactly, it's from a paper. You can see the paper in `/data/`.

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

pd.set_option("display.max_columns", 200)
pd.set_option("display.max_colwidth", 200)
%matplotlib inline

# Reading in our data

## Every single person in a crash

We'll start by reading in the list of people who were involved in an accident.

**Call this dataframe `people`.**

In [2]:
people = pd.read_csv('data/combined-person-data.csv')
people.head()


# 01	No Injury
# 02	Non-incapacitating Injury
# 03	Possible Incapacitating Injury
# 04	Incapacitating/Disabled Injury
# 05	Fatal Injury


C:\Users\Dominik Balmer\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (5,15,26) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


AIRBAG_DEPLOYED  ALCOHOL_TESTTYPE_CODE  ALCOHOL_TEST_CODE  BAC_CODE  \
0              1.0                    NaN                0.0       NaN   
1              1.0                    NaN                0.0       NaN   
2              1.0                    NaN                0.0       NaN   
3              1.0                    NaN                0.0       NaN   
4              1.0                    NaN                0.0       NaN   

  CDL_FLAG CLASS  CONDITION_CODE        DATE_OF_BIRTH DRUG_TESTRESULT_CODE  \
0        N     C             0.0  1952-04-20 00:00:00                  NaN   
1        N   NaN             0.0  1985-05-28 00:00:00                  NaN   
2        N     C             0.0  1960-10-04 00:00:00                  NaN   
3        N     B             0.0  1971-05-28 00:00:00                  NaN   
4        N     D             0.0  1955-04-23 00:00:00                  NaN   

   DRUG_TEST_CODE  EJECT_CODE EMS_UNIT_LABEL  EQUIP_PROB_CODE FAULT_FLAG  \
0             0.0         1.0            NaN              1.0          N   
1             0.0         0.0            NaN              0.0          N   
2             0.0         0.0            NaN              0.0          Y   
3             0.0         1.0            NaN              0.0          N   
4             0.0         0.0            NaN              1.0          N   

   INJ_SEVER_CODE LICENSE_STATE_CODE  MOVEMENT_CODE  OCC_SEAT_POS_CODE  \
0               1                 PA            NaN                NaN   
1               1                 MD            NaN                NaN   
2               1                 MD            NaN                NaN   
3               1                 MD            NaN                NaN   
4               1                 SC            NaN                NaN   

   PED_LOCATION_CODE  PED_OBEY_CODE  PED_TYPE_CODE  PED_VISIBLE_CODE  \
0                NaN            NaN            NaN               NaN   
1                NaN            NaN            NaN               NaN   
2                NaN            NaN            NaN               NaN   
3                NaN            NaN            NaN               NaN   
4                NaN            NaN            NaN               NaN   

                              PERSON_ID PERSON_TYPE   REPORT_NO  \
0  48dd00ee-e033-47e7-ad1e-0b734020301b           D  AB4284000S   
1  166296bd-ffd3-4c16-aa74-4f4bf4139d8d           D  AB4313000X   
2  f3b2743f-fbc3-4345-9419-56a0ca29102c           D  AB4313000X   
3  5bbe589b-a2db-4dbb-be9b-17a800d69a08           D  AB4313000Y   
4  b914136f-5ecd-46bb-94ec-ff5d4136c3eb           D  AB4669001F   

   SAF_EQUIP_CODE SEX_CODE                            VEHICLE_ID  
0            13.0        F  eb6aadb8-dacb-4744-a1a7-ab812c96f27f  
1            13.0        F  b463eb20-2f01-4200-9d6f-b18888ce2593  
2            13.0        F  3c8629d0-d524-47c1-bfbc-b18e07f3087e  
3             0.0        F  c4628cdb-f295-4a24-8a4b-653741ac6ae7  
4            13.0        F  cdda1580-fd79-4358-8819-c2250f494591

How often did each severity of injury show up? (e.g. not injured, non-incapacitating injury, etc)

In [3]:
people.INJ_SEVER_CODE.value_counts()

1    716806
2     82642
3     76801
4     10353
5      1681
Name: INJ_SEVER_CODE, dtype: int64

We're only interested in fatalities, so let's create a new `is_fatality` column for when people received a fatal injury.



**Confirm there were 1681 people with fatal injuries.**

In [4]:
people['is_fatality'] = people.INJ_SEVER_CODE.replace({5: 1, 4: 0, 3:0, 2:0, 1:0})
people.head() 
people.is_fatality.value_counts()

0    886602
1      1681
Name: is_fatality, dtype: int64

## Working on Features

### Starting our analysis

We're going to run a regression on the impact of being **male vs female on crash fatalities**. Prepare a dataframe called `train_df` with the appropriate information in it.

* **Tip:** What column(s) are your input, and what is your output? Aka independent and dependent variables
* **Tip:** You'll need to convert your input column into something numeric, I suggest using `.replace`
* **Tip:** We aren't interested in the "Unknown" sex - either filtering or `np.nan` + `.dropna()` might be useful ways to get rid of those columns

F	Female
M	Male
U	Unkown


In [5]:
train_df = people

train_df['genders'] = train_df.SEX_CODE.replace({'F': 1, 'M': 0, 'U': np.nan})
train_df = train_df.dropna(subset=['genders'])
train_df['genders'].value_counts()

0.0    460973
1.0    354854
Name: genders, dtype: int64

In [6]:
train_df.genders.value_counts()

0.0    460973
1.0    354854
Name: genders, dtype: int64

Confirm that your `train_df` has two columns and 815,827 rows.

> **Tip:** If you have more rows, make sure you dropped all of the rows with Unknown sex.
>
> **Tip:** If you have more columns, make sure you only have your input and output columns.

In [7]:
train_df.shape

(815827, 30)

In [8]:
train_df = train_df[['is_fatality', 'genders']]

In [9]:
train_df.shape

(815827, 2)

### Run your regression

See the effect of sex on whether the person's injuries are fatal or not. I want to see a result dataframe that includes:

* Feature name
* Coefficient
* Odds ratio

In [10]:
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality

In [11]:
clf = LogisticRegression(C=1e9, solver='lbfgs')

clf.fit(X, y)

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [12]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients).round(4)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
0  genders                      -0.71481      0.4893

### Use words to interpret this result

In [13]:
# women are almost 50% less likely to be injured in an accident 

## Adding more features

The actual crash data has more details - whether it was snowy/icy, whether it was a highway, etc. 

Read in `combined-crash-data.csv` and merge it with our people dataset. I'll save you a lookup: the `REPORT_NO` is what matches between the two.

In [14]:
crash = pd.read_csv('data/combined-crash-data.csv')
crash.head()

C:\Users\Dominik Balmer\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (27,30,33) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


ACC_DATE  ACC_TIME AGENCY_CODE AREA_CODE  COLLISION_TYPE_CODE  \
0  2018-04-10 00:00:00  01:50:00         MSP       UNK                   17   
1  2018-05-02 00:00:00  11:06:00      ELKTON       UNK                    5   
2  2018-05-02 00:00:00  11:54:00  WASHINGTON       UNK                   88   
3  2018-05-02 00:00:00  12:45:00  MONTGOMERY       UNK                    7   
4  2018-05-02 00:00:00  12:02:00         MSP       UNK                   11   

   COUNTY_NO C_M_ZONE_FLAG  DISTANCE DISTANCE_DIR_FLAG FEET_MILES_FLAG  \
0       18.0             N      0.00                 N               F   
1        7.0             N     15.00                 S               F   
2       21.0             N      1.00                 S               M   
3       15.0             N      0.00                 E               F   
4        8.0             N      0.15                 N               M   

   FIX_OBJ_CODE  HARM_EVENT_CODE1  HARM_EVENT_CODE2  JUNCTION_CODE  LANE_CODE  \
0         22.03              16.0              11.0           1.00        NaN   
1          0.00               1.0               1.0           3.00        NaN   
2         10.00              16.0               9.0           0.00        1.0   
3          0.00               1.0               1.0           2.00        2.0   
4          0.00               1.0               1.0           9.04        2.0   

    LATITUDE  LIGHT_CODE LOC_CODE LOGMILE_DIR_FLAG  LOG_MILE  LONGITUDE  \
0  38.277230         4.0      NaN                N      2.09 -76.682876   
1  39.613747         1.0      NaN                N     19.41 -75.837454   
2  39.680144         1.0      NaN                N      2.60 -77.514781   
3  39.015649         1.0    55122                E      0.00 -77.042728   
4  38.620875         1.0      NaN                N      1.02 -76.901087   

      MAINROAD_NAME  MUNI_CODE  RD_COND_CODE  RD_DIV_CODE  REFERENCE_NO  \
0  NEWTOWNE NECK RD        0.0           1.0          1.0         166.0   
1       N BRIDGE ST       52.0           1.0          3.0         362.0   
2        RITCHIE RD        0.0           1.0          1.0         270.0   
3    FOREST GLEN RD        0.0           0.0          4.0          97.0   
4    LEONARDTOWN RD        0.0           1.0          3.0         644.0   

  REFERENCE_ROAD_NAME REFERENCE_SUFFIX REFERENCE_TYPE_CODE    REPORT_NO  \
0         ROSEBANK RD              NaN                  CO  MSP6188002Q   
1           LAUREL DR              NaN                  MU   BK0227001M   
2             WISE RD              NaN                  CO   ZU8005001W   
3         GEORGIA AVE              NaN                  MD  MCP2891005S   
4      POST OFFICE RD              NaN                  CO  MSP6743004Q   

             REPORT_TYPE ROUTE_TYPE_CODE  RTE_NO RTE_SUFFIX SIGNAL_FLAG  \
0           Injury Crash              MD   243.0        NaN           N   
1           Injury Crash              MD   213.0        NaN           N   
2           Injury Crash              CO   269.0        NaN           N   
3  Property Damage Crash              CO   697.0        NaN           Y   
4  Property Damage Crash              MD     5.0         BU           N   

   SURF_COND_CODE  WEATHER_CODE  
0             2.0          6.01  
1             2.0          6.01  
2             2.0          6.01  
3             2.0          6.01  
4             2.0          6.01

In [15]:
merged = pd.merge(crash, people, on='REPORT_NO')
merged.head()

ACC_DATE  ACC_TIME AGENCY_CODE AREA_CODE  COLLISION_TYPE_CODE  \
0  2018-04-10 00:00:00  01:50:00         MSP       UNK                   17   
1  2018-05-02 00:00:00  11:06:00      ELKTON       UNK                    5   
2  2018-05-02 00:00:00  11:06:00      ELKTON       UNK                    5   
3  2018-05-02 00:00:00  11:06:00      ELKTON       UNK                    5   
4  2018-05-02 00:00:00  11:06:00      ELKTON       UNK                    5   

   COUNTY_NO C_M_ZONE_FLAG  DISTANCE DISTANCE_DIR_FLAG FEET_MILES_FLAG  \
0       18.0             N       0.0                 N               F   
1        7.0             N      15.0                 S               F   
2        7.0             N      15.0                 S               F   
3        7.0             N      15.0                 S               F   
4        7.0             N      15.0                 S               F   

   FIX_OBJ_CODE  HARM_EVENT_CODE1  HARM_EVENT_CODE2  JUNCTION_CODE  LANE_CODE  \
0         22.03              16.0              11.0            1.0        NaN   
1          0.00               1.0               1.0            3.0        NaN   
2          0.00               1.0               1.0            3.0        NaN   
3          0.00               1.0               1.0            3.0        NaN   
4          0.00               1.0               1.0            3.0        NaN   

    LATITUDE  LIGHT_CODE LOC_CODE LOGMILE_DIR_FLAG  LOG_MILE  LONGITUDE  \
0  38.277230         4.0      NaN                N      2.09 -76.682876   
1  39.613747         1.0      NaN                N     19.41 -75.837454   
2  39.613747         1.0      NaN                N     19.41 -75.837454   
3  39.613747         1.0      NaN                N     19.41 -75.837454   
4  39.613747         1.0      NaN                N     19.41 -75.837454   

      MAINROAD_NAME  MUNI_CODE  RD_COND_CODE  RD_DIV_CODE  REFERENCE_NO  \
0  NEWTOWNE NECK RD        0.0           1.0          1.0         166.0   
1       N BRIDGE ST       52.0           1.0          3.0         362.0   
2       N BRIDGE ST       52.0           1.0          3.0         362.0   
3       N BRIDGE ST       52.0           1.0          3.0         362.0   
4       N BRIDGE ST       52.0           1.0          3.0         362.0   

  REFERENCE_ROAD_NAME REFERENCE_SUFFIX REFERENCE_TYPE_CODE    REPORT_NO  \
0         ROSEBANK RD              NaN                  CO  MSP6188002Q   
1           LAUREL DR              NaN                  MU   BK0227001M   
2           LAUREL DR              NaN                  MU   BK0227001M   
3           LAUREL DR              NaN                  MU   BK0227001M   
4           LAUREL DR              NaN                  MU   BK0227001M   

    REPORT_TYPE ROUTE_TYPE_CODE  RTE_NO RTE_SUFFIX SIGNAL_FLAG  \
0  Injury Crash              MD   243.0        NaN           N   
1  Injury Crash              MD   213.0        NaN           N   
2  Injury Crash              MD   213.0        NaN           N   
3  Injury Crash              MD   213.0        NaN           N   
4  Injury Crash              MD   213.0        NaN           N   

   SURF_COND_CODE  WEATHER_CODE  AIRBAG_DEPLOYED  ALCOHOL_TESTTYPE_CODE  \
0             2.0          6.01              4.0                    2.0   
1             2.0          6.01              1.0                    NaN   
2             2.0          6.01              1.0                    NaN   
3             2.0          6.01              1.0                    NaN   
4             2.0          6.01              5.0                    NaN   

   ALCOHOL_TEST_CODE  BAC_CODE CDL_FLAG CLASS  CONDITION_CODE  \
0                3.0       NaN        N     C             2.0   
1                NaN       NaN      NaN   NaN             NaN   
2                0.0       NaN        N     A             1.0   
3                0.0       NaN        N     C             1.0   
4                0.0       NaN        N     B            99.0   

         DATE_O

### Examining more possible features

How often was it wet, dry, snowy, icy, etc? **What was the most common condition?**

In [16]:
merged.SURF_COND_CODE.value_counts()

2.00     619569
1.00     147803
0.00      30921
3.00       9418
4.00       9111
99.00      3616
6.03       1697
88.00      1064
5.00        447
7.01        258
9.88        141
8.05         36
Name: SURF_COND_CODE, dtype: int64

Do you feel that a **Dry** road condition should be the average of **Wet** and **Snow?**

In [17]:
merged['condition'] = merged.SURF_COND_CODE.replace({2.00 : 'Dry', 1.00 : 'Wet', 0.00 : 'Not Applicable', 4.00 : 'Ice', 3.00 : 'Snow', 6.03 : 'Slush', 99.00 : 'Unknown', 88.00 : 'Other', 5.00 : 'Mud, Dirt, Gravel', 7.01 : 'Water (standing/moving)', 9.88 : 'Oil', 8.05 : 'Sand'})

The answer to that should be *no*, which means we can't use this data as numeric data. We want a different coefficient for each of these - I want to know the impact of dry, the impact of wet, the impact of snow, all separately.

Start by **replacing each code with a proper description.** I'll even include them here:

* `00` - Not Applicable
* `01`	- Wet
* `02`	- Dry
* `03`	- Snow
* `04`	- Ice
* `05`	- Mud, Dirt, Gravel
* `06`	- Slush
* `07`	- Water (standing/moving)
* `08`	- Sand
* `09`	- Oil
* `88`	- Other
* `99`	- Unknown

But watch out, pandas read the column in as numbers so they might have come through a little differently than their codes.

Confirm you have 147,803 wet, and a few codes you can't understand, like `6.03` and `7.01`.

In [18]:
merged['condition'].value_counts()

Dry                        619569
Wet                        147803
Not Applicable              30921
Snow                         9418
Ice                          9111
Unknown                      3616
Slush                        1697
Other                        1064
Mud, Dirt, Gravel             447
Water (standing/moving)       258
Oil                           141
Sand                           36
Name: condition, dtype: int64

Replace the codes you don't understand with `Other`.

In [19]:
merged['condition'] = merged.condition.replace({'Slush': 'Other', 'Water (standing/moving)' : 'Other', 'Sand' : 'Other', 'Oil' : 'Other'})

Confirm you have 3,196 'Other'.

In [20]:
merged.condition.value_counts()

Dry                  619569
Wet                  147803
Not Applicable        30921
Snow                   9418
Ice                    9111
Unknown                3616
Other                  3196
Mud, Dirt, Gravel       447
Name: condition, dtype: int64

## One-hot encoding

We're going to use `pd.get_dummies` to build a variable you'll call `surf_dummies`. Each surface condition should be a `0` or `1` as to whether it was that condition (dry, icy, wet, etc).

Use a `prefix=` so we know they are **surface** conditions.

You'll want to drop the column you'll use as the reference category.

**Before we do this: which column works best as the reference?**

In [21]:
pd.get_dummies(merged.condition, prefix='surface').head()

surface_Dry  surface_Ice  surface_Mud, Dirt, Gravel  \
0            1            0                          0   
1            1            0                          0   
2            1            0                          0   
3            1            0                          0   
4            1            0                          0   

   surface_Not Applicable  surface_Other  surface_Snow  surface_Unknown  \
0                       0              0             0                0   
1                       0              0             0                0   
2                       0              0             0                0   
3                       0              0             0                0   
4                       0              0             0                0   

   surface_Wet  
0            0  
1            0  
2            0  
3            0  
4            0

Now build your `surf_dummies` variable.

In [22]:
surf_dummies = pd.get_dummies(merged.condition, prefix='surface').drop('surface_Dry', axis=1)
surf_dummies.sum()

surface_Ice                    9111
surface_Mud, Dirt, Gravel       447
surface_Not Applicable        30921
surface_Other                  3196
surface_Snow                   9418
surface_Unknown                3616
surface_Wet                  147803
dtype: int64

Confirm your `surf_dummies` looks roughly like this:

|surface_Ice|Surce_Mud, Dirt, Gravel|surface_Not Applicable|...|surface_Wet|
|---|---|---|---|---|
|0|0|0|...|0|
|0|0|0|...|0|
|0|0|1|...|0|
|0|0|1|...|0|
|0|0|0|...|1|

## Another regression

Let's run another regression to see the impact of both **sex and surface condition** on fatalities.

### Build your `train_df`

To build your `train_df`, I recommend doing it either of these two ways:

```python
train_df = pd.DataFrame({
    'col1': merged.col1,
    'col2': merged.col2,
    'col3': merged.col3,
})
train_df = train_df.join(surf_dummies)
train_df = train_df.dropna()
```

or like this:

```python
train_df = train_df[['col1','col2','col3']].copy()
train_df = train_df.join(surf_dummies)
train_df = train_df.dropna()
```

The second one is shorter, but the first one makes it easier to use comments to remove columns later.


In [23]:
train_df = pd.DataFrame({
    'is_fatality': merged['is_fatality'],
    'genders': merged['genders'],
    'condition': merged['condition'],
})

train_df = merged[['is_fatality', "condition"]].join(surf_dummies)
train_df = train_df.dropna()

In [24]:
train_df = merged[['is_fatality', "genders"]].join(surf_dummies)
train_df = train_df.dropna()

### Run your regression and check your odds ratios

Actually no, wait, first - what kind of surface do you think will have the **highest fatality rate?**

In [25]:
train_df.dtypes

is_fatality                    int64
genders                      float64
surface_Ice                    uint8
surface_Mud, Dirt, Gravel      uint8
surface_Not Applicable         uint8
surface_Other                  uint8
surface_Snow                   uint8
surface_Unknown                uint8
surface_Wet                    uint8
dtype: object

In [26]:
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality

In [27]:
clf = LogisticRegression(C=1e9, solver='lbfgs')
clf.fit(X, y)

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [28]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients).round(4)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
2  surface_Mud, Dirt, Gravel                      1.115150      3.0500
4              surface_Other                      0.596230      1.8153
6            surface_Unknown                      0.340200      1.4052
7                surface_Wet                     -0.098825      0.9059
1                surface_Ice                     -0.557708      0.5725
5               surface_Snow                     -0.652480      0.5208
0                    genders                     -0.718133      0.4877
3     surface_Not Applicable                     -0.730582      0.4816

Confirm your `train_df` has 815,843 rows and 9 columns.

* **Tip:** When you run your regression, if you get an error about not knowing what to do with `U`, it's because you didn't convert your sex to numbers (or if you did, you didn't do it in your original dataframe)

In [29]:
train_df.shape

(815843, 9)

**Is this what you expected?** Why do you think this result might be the case?

# More features: Vehicles

Maybe the car they're in is related to the car they were in. Luckily, we have this information - **read in `combined_vehicle_data` as `vehicles`.**

In [30]:
vehicles1 = pd.read_csv('data/combined-vehicle-data.csv')

## Weights of those cars

The car weights are stored in **another file** since the info had to come from an API. I looked up the VINs - vehicle identification numbers - in a government database to try to get data for each of them.

**Read them and build a new dataframe that is both the vehicle data along with their weights.** You can call it `vehicles` since you don't need the original weightless vehicle data any more.

In [31]:
vins_and_weights = pd.read_csv('data/vins_and_weights.csv')

In [32]:
vehicles = pd.merge(vehicles1, vins_and_weights, left_on='VIN_NO', right_on='VIN')

Confirm that your combined `vehicles` dataset should have 534,436 rows and 35 columns. And yes, that's less than we were working with before - you haven't combined it with the people/crashes dataset yet.

In [33]:
vehicles.shape

(534436, 35)

### Filter your data

We only want vehicles that are "normal" - somewhere between 1500 and 6000 pounds. Filter your vehicles to only include those in that weight range.

In [34]:
vehicles = vehicles[(vehicles['weight'] > 1500) & (vehicles['weight'] < 6000)]

Confirm that you have 532,370 vehicles in the dataset.

In [35]:
vehicles.shape

(532370, 35)

## Add this vehicle information to your merged data

Now we'll have a dataframe that contains information on:

* The people themselves and their injuries
* The crash
* The vehicles

Every person came with a `VEHICLE_ID` column that is the vehicle they were in. You'll want to merge on that.

In [36]:
merged2 = pd.merge(merged, vehicles, on='VEHICLE_ID')

Confirm you have 99 columns and 616,212 rows. **That is a lot of possible features!**

In [37]:
merged2.shape

(616212, 101)

## Another regression, because we can't get enough

Build another `train_df` and run another regression about **how car weight impacts the chance of fatalities**. You'll want to confirm that your dataset has 616,212 and 2 columns.

In [38]:
train_df = merged2.copy()
train_df = train_df[['is_fatality', 'weight']] 

In [39]:
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality 

In [40]:
clf = LogisticRegression(C=1e9, solver='lbfgs')
clf.fit(X, y)

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [41]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False) 

feature  coefficient (log odds ratio)  odds ratio
0  weight                     -0.000147    0.999853

**Can you translate that into plain English?** Remember weight is in **pounds**.

I feel like pounds isn't the best measure for something like this. Remember how we had to adjust percentages with AP and life expecntancy, and then change around the way we said things? It sounded like this:

> Every 10% increase in unemployment translates to a year and a half loss of life expectancy

Instead of every single pound, maybe we could do every... some other number of pounds? One hundred? One thousand?

**Run another regression with weight in thousands of pounds.** Get another odds ratio. Give me another sentence English.

In [42]:
train_df['weight'] = train_df.weight/1000 

In [43]:
train_df = merged2.copy()
train_df = train_df[['is_fatality', 'weight']]
train_df['weight'] = train_df.weight/1000 

In [44]:
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality 

In [45]:
clf = LogisticRegression(C=1e9, solver='lbfgs')
clf.fit(X, y)

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [46]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False) 

feature  coefficient (log odds ratio)  odds ratio
0  weight                     -0.146881    0.863397

In [47]:
# Every thousand pounds heavier your car is increase translates to a 15% decrease in fatalities

# Two-car accidents, struck and striker

Here's the thing, though: **it isn't just the weight of your car.** It's the weight of both cars! If I'm in a big car and I have a wreck with a smaller car, it's the smaller car that's in trouble.

To get that value, we need to do some **feature engineering**, some calculating of *new* variables from our *existing* variables.

We need to jump through some hoops to do that.

##  Two-car accidents

First we're going to count how many vehicles were in each accident. Since we're looking to compare the weight of two cars hitting each other, **we're only going to want crashes with only two cars.**

In [48]:
counted = vehicles.REPORT_NO.value_counts()
counted.head(10)

CB53480003      28
MDTA1229000H    26
AS0286000B      14
CE4636002C      14
MSP67460095     13
MSP6063006C     13
CC0237003W      12
MSP63070049     12
AE5607002W      12
MSP6297005S     11
Name: REPORT_NO, dtype: int64

By using `.value_counts` I can see how many cars were in each crash, and now I'm going to filter to get a list of all of the ones with two vehicles.

In [49]:
two_car_report_nos = counted[counted == 2].index
two_car_report_nos

Index(['MCP23920006', 'MSP65060018', 'AC2112000N', 'BX5083001X', 'BX5034000G',
       'MSP63540027', 'MSP5682001F', 'DD55980028', 'BS0532000B', 'DA4104000T',
       ...
       'DA3772002Q', 'ADI804004N', 'AE48830033', 'MCP21910015', 'DA2910001K',
       'ADH6080004', 'DM84620005', 'BX5083000V', 'CB5131001T', 'ADJ5300027'],
      dtype='object', length=137214)

And now we'll filter my vehicles so we only have those that were in two-vehicle crashes.

In [50]:
vehicles = vehicles[vehicles.REPORT_NO.isin(two_car_report_nos)]

### Struck and striker

To do the math correctly, we need both the risk of someone dying in the smaller car _and_ the risk of someone dying in the bigger car. To do this we need to separate our cars into two groups:

* The 'struck' vehicle: did the person die inside?
* The 'striker' vehicle: how much heavier was it than the struck car?

But we don't know which car was which, so we have to try out both versions - pretending car A was the striker, then pretending car B was the striker. It's hard to explain, but you can read `Pounds That Kill - The External Costs of Vehicle Weight.pdf` for more details on how it works.

In [51]:
cars_1 = vehicles.drop_duplicates(subset='REPORT_NO', keep='first')
cars_2 = vehicles.drop_duplicates(subset='REPORT_NO', keep='last')

In [52]:
cars_merged_1 = cars_1.merge(cars_2, on='REPORT_NO', suffixes=['_striker', '_struck'])
cars_merged_2 = cars_2.merge(cars_1, on='REPORT_NO', suffixes=['_striker', '_struck'])
vehicles_complete = pd.concat([cars_merged_1, cars_merged_2])
vehicles_complete.head()

AREA_DAMAGED_CODE1_striker  AREA_DAMAGED_CODE2_striker  \
0                         6.0                         7.0   
1                         8.0                         NaN   
2                        12.0                         NaN   
3                         1.0                        11.0   
4                         2.0                         NaN   

   AREA_DAMAGED_CODE3_striker  AREA_DAMAGED_CODE_IMP1_striker  \
0                         NaN                             6.0   
1                         NaN                             8.0   
2                         NaN                            12.0   
3                        12.0                            12.0   
4                         NaN                             2.0   

   AREA_DAMAGED_CODE_MAIN_striker  BODY_TYPE_CODE_striker  \
0                             6.0                    20.0   
1                             8.0                     2.0   
2                            12.0                     2.0   
3                            12.0                     2.0   
4                             2.0                     2.0   

  COMMERCIAL_FLAG_striker CONTI_DIRECTION_CODE_striker  \
0                       N                            N   
1                       N                            W   
2                       N                            N   
3                       N                            S   
4                       N                            S   

   CV_BODY_TYPE_CODE_striker  DAMAGE_CODE_striker DRIVERLESS_FLAG_striker  \
0                        NaN                    3                       N   
1                        NaN                    2                       N   
2                        NaN                    4                       N   
3                        NaN                    4                       N   
4                        NaN                   99                       N   

  FIRE_FLAG_striker GOING_DIRECTION_CODE_striker  GVW_CODE_striker  \
0                 N                            N               NaN   
1               NaN                            W               NaN   
2               NaN                            N               NaN   
3                 N                            S               NaN   
4                 N                            S               NaN   

   HARM_EVENT_CODE_striker HAZMAT_SPILL_FLAG_striker HIT_AND_RUN_FLAG_striker  \
0                      1.0                       NaN                        N   
1                      1.0                       NaN                        N   
2                      0.0                       NaN                        N   
3                      1.0                       NaN                        N   
4                      1.0                       NaN                        Y   

  HZM_NUM_striker  MOVEMENT_CODE_striker  NUM_AXLES_striker  \
0             NaN                    6.0                NaN   
1             NaN                    8.0                NaN   
2             NaN                    1.0                NaN   
3             NaN                   12.0                NaN   
4             NaN                    1.0                NaN   

  PARKED_FLAG_striker   REPORT_NO  SPEED_LIMIT_striker  \
0                   N  CB5190006B                   55   
1                   N  ADJ4590035                    5   
2                   N  ADJ849000Z                   10   
3                   N  AE5207008Z                   30   
4                   N  ADJ859000Y                   15   

  TOWED_AWAY_FLAG_striker  TOWED_VEHICLE_CONFIG_CODE_striker  \
0                       N                                  0   
1                       N                                  0   
2                       N                                  0   
3                       Y                                  0   
4                       N                                  0   

                     VEHICLE_ID_striker VEH_M

## Put people in their cars

Which car was each person in? We'll assign that now.

In [53]:
merged = people.merge(vehicles_complete, left_on='VEHICLE_ID', right_on='VEHICLE_ID_struck')
merged.head(3)

AIRBAG_DEPLOYED  ALCOHOL_TESTTYPE_CODE  ALCOHOL_TEST_CODE  BAC_CODE  \
0              1.0                    NaN                0.0       NaN   
1              1.0                    NaN                0.0       NaN   
2              1.0                    NaN                0.0       NaN   

  CDL_FLAG CLASS  CONDITION_CODE        DATE_OF_BIRTH DRUG_TESTRESULT_CODE  \
0        N   NaN             0.0  1985-05-28 00:00:00                  NaN   
1        N     C             0.0  1960-10-04 00:00:00                  NaN   
2        N     C             0.0  1990-11-05 00:00:00                  NaN   

   DRUG_TEST_CODE  EJECT_CODE EMS_UNIT_LABEL  EQUIP_PROB_CODE FAULT_FLAG  \
0             0.0         0.0            NaN              0.0          N   
1             0.0         0.0            NaN              0.0          Y   
2             0.0         1.0            NaN              1.0          N   

   INJ_SEVER_CODE LICENSE_STATE_CODE  MOVEMENT_CODE  OCC_SEAT_POS_CODE  \
0               1                 MD            NaN                NaN   
1               1                 MD            NaN                NaN   
2               1                 MD            NaN                NaN   

   PED_LOCATION_CODE  PED_OBEY_CODE  PED_TYPE_CODE  PED_VISIBLE_CODE  \
0                NaN            NaN            NaN               NaN   
1                NaN            NaN            NaN               NaN   
2                NaN            NaN            NaN               NaN   

                              PERSON_ID PERSON_TYPE REPORT_NO_x  \
0  166296bd-ffd3-4c16-aa74-4f4bf4139d8d           D  AB4313000X   
1  f3b2743f-fbc3-4345-9419-56a0ca29102c           D  AB4313000X   
2  21fff0f9-ed13-4b87-ac04-29d018aff9d4           D  AB5218001Y   

   SAF_EQUIP_CODE SEX_CODE                            VEHICLE_ID  is_fatality  \
0            13.0        F  b463eb20-2f01-4200-9d6f-b18888ce2593            0   
1            13.0        F  3c8629d0-d524-47c1-bfbc-b18e07f3087e            0   
2            13.0        F  4dea42c3-e02c-4c6b-8ea0-c2f8c17f147f            0   

   genders  AREA_DAMAGED_CODE1_striker  AREA_DAMAGED_CODE2_striker  \
0      1.0                         3.0                         NaN   
1      1.0                         7.0                         NaN   
2      1.0                        10.0                        11.0   

   AREA_DAMAGED_CODE3_striker  AREA_DAMAGED_CODE_IMP1_striker  \
0                         NaN                             3.0   
1                         NaN                             7.0   
2                        12.0                            11.0   

   AREA_DAMAGED_CODE_MAIN_striker  BODY_TYPE_CODE_striker  \
0                             3.0                     2.0   
1                             7.0                     2.0   
2                            11.0                     2.0   

  COMMERCIAL_FLAG_striker CONTI_DIRECTION_CODE_striker  \
0                       N                            E   
1                       N                            S   
2                       N                            N   

   CV_BODY_TYPE_CODE_striker  DAMAGE_CODE_striker DRIVERLESS_FLAG_striker  \
0                        NaN                    2                       N   
1                        NaN                    2                       N   
2                        NaN                    3                       N   

  FIRE_FLAG_striker GOING_DIRECTION_CODE_striker  GVW_CODE_striker  \
0               NaN                            E               NaN   
1                 N                            S               NaN   
2                 N                            N               NaN   

   HARM_EVENT_CODE_striker HAZMAT_SPILL_FLAG_striker HIT_AND_RUN_FLAG_striker  \
0                      1.0                       NaN                        Y   
1                      1.0                       NaN                        N   
2                      1.0                       NaN   

# Add the crash details

You did this already! I'm going to do it for you. We're merging on `REPORT_NO_x` because there are so many `REPORT_NO` columns duplicated across our files that pandas started giving them weird names.

In [54]:
merged = merged.merge(crash, left_on='REPORT_NO_x', right_on='REPORT_NO')
merged.head(3)

AIRBAG_DEPLOYED  ALCOHOL_TESTTYPE_CODE  ALCOHOL_TEST_CODE  BAC_CODE  \
0              1.0                    NaN                0.0       NaN   
1              1.0                    NaN                0.0       NaN   
2              1.0                    NaN                0.0       NaN   

  CDL_FLAG CLASS  CONDITION_CODE        DATE_OF_BIRTH DRUG_TESTRESULT_CODE  \
0        N   NaN             0.0  1985-05-28 00:00:00                  NaN   
1        N     C             0.0  1960-10-04 00:00:00                  NaN   
2        N     C             0.0  1990-11-05 00:00:00                  NaN   

   DRUG_TEST_CODE  EJECT_CODE EMS_UNIT_LABEL  EQUIP_PROB_CODE FAULT_FLAG  \
0             0.0         0.0            NaN              0.0          N   
1             0.0         0.0            NaN              0.0          Y   
2             0.0         1.0            NaN              1.0          N   

   INJ_SEVER_CODE LICENSE_STATE_CODE  MOVEMENT_CODE  OCC_SEAT_POS_CODE  \
0               1                 MD            NaN                NaN   
1               1                 MD            NaN                NaN   
2               1                 MD            NaN                NaN   

   PED_LOCATION_CODE  PED_OBEY_CODE  PED_TYPE_CODE  PED_VISIBLE_CODE  \
0                NaN            NaN            NaN               NaN   
1                NaN            NaN            NaN               NaN   
2                NaN            NaN            NaN               NaN   

                              PERSON_ID PERSON_TYPE REPORT_NO_x  \
0  166296bd-ffd3-4c16-aa74-4f4bf4139d8d           D  AB4313000X   
1  f3b2743f-fbc3-4345-9419-56a0ca29102c           D  AB4313000X   
2  21fff0f9-ed13-4b87-ac04-29d018aff9d4           D  AB5218001Y   

   SAF_EQUIP_CODE SEX_CODE                            VEHICLE_ID  is_fatality  \
0            13.0        F  b463eb20-2f01-4200-9d6f-b18888ce2593            0   
1            13.0        F  3c8629d0-d524-47c1-bfbc-b18e07f3087e            0   
2            13.0        F  4dea42c3-e02c-4c6b-8ea0-c2f8c17f147f            0   

   genders  AREA_DAMAGED_CODE1_striker  AREA_DAMAGED_CODE2_striker  \
0      1.0                         3.0                         NaN   
1      1.0                         7.0                         NaN   
2      1.0                        10.0                        11.0   

   AREA_DAMAGED_CODE3_striker  AREA_DAMAGED_CODE_IMP1_striker  \
0                         NaN                             3.0   
1                         NaN                             7.0   
2                        12.0                            11.0   

   AREA_DAMAGED_CODE_MAIN_striker  BODY_TYPE_CODE_striker  \
0                             3.0                     2.0   
1                             7.0                     2.0   
2                            11.0                     2.0   

  COMMERCIAL_FLAG_striker CONTI_DIRECTION_CODE_striker  \
0                       N                            E   
1                       N                            S   
2                       N                            N   

   CV_BODY_TYPE_CODE_striker  DAMAGE_CODE_striker DRIVERLESS_FLAG_striker  \
0                        NaN                    2                       N   
1                        NaN                    2                       N   
2                        NaN                    3                       N   

  FIRE_FLAG_striker GOING_DIRECTION_CODE_striker  GVW_CODE_striker  \
0               NaN                            E               NaN   
1                 N                            S               NaN   
2                 N                            N               NaN   

   HARM_EVENT_CODE_striker HAZMAT_SPILL_FLAG_striker HIT_AND_RUN_FLAG_striker  \
0                      1.0                       NaN                        Y   
1                      1.0                       NaN                        N   
2                      1.0                       NaN   

## Filter

We already filtered out vehicles by weight, so we don't have to do that again.

# Calculated features

I'm sure you forgot what all the features are, so we'll bring back whether there was a fatality or not

## Feature: Accident was fatal

In [55]:
merged['had_fatality'] = (merged.INJ_SEVER_CODE == 5).astype(int)
merged.had_fatality.value_counts()

0    334152
1       244
Name: had_fatality, dtype: int64

## Feature: Weight difference

**Remove everything missing weights for strikers or struck vehicles.** You might need to `merged.columns` to remind yourself what the column names are.

In [56]:
merged = merged.dropna(subset=['weight_striker', 'weight_struck'])

Confirm your dataset has 334,396 rows.

In [57]:
merged.shape

(334396, 137)

Create a new feature called `weight_diff` about how much heavier the striking car was compared to the struck car. **Make sure you've done the math correctly!**

In [58]:
merged['weight_diff'] = merged.weight_striker - merged.weight_struck

### Feature adjustment

Make all of your weight columns in **thousands of pounds** instead of just in pounds. It'll help you interpret your results much better.

In [59]:
merged['weight_diff'] = merged['weight_diff'] / 1000
merged['weight_struck'] = merged['weight_struck'] / 1000
merged['weight_striker'] = merged['weight_striker'] / 1000

# Another regression!!!

**What is the impact of weight difference on fatality rate?** Create your `train_df`, drop missing values, run your regression, analyze your odds ratios.

In [60]:
train_df = merged[['is_fatality', 'weight_diff']]
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality
logit = LogisticRegression()
logit.fit(X, y)

C:\Users\Dominik Balmer\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [61]:
feature_names = X.columns
coefficients = clf.coef_[0]
pd.DataFrame({
   'feature': feature_names,
   'coefficient (log odds ratio)': coefficients,
   'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
0  weight_diff                     -0.146881    0.863397

Please translate your odds ratio into plain English.

## Adding in more features

How about speed limit? That's important, right? We can add the speed limit of the striking vehicle with `SPEED_LIMIT_striker`.

In [62]:
train_df = merged[['is_fatality', 'weight_diff', 'SPEED_LIMIT_striker']]

In [63]:
X = train_df.drop(columns=['is_fatality'])
y = train_df.is_fatality
logit = LogisticRegression()
logit.fit(X, y)

C:\Users\Dominik Balmer\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [69]:
feature_names = X.columns
coefficients = logit.coef_[0]
pd.DataFrame({
   'feature': feature_names,
   'coefficient (log odds ratio)': coefficients,
   'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
0          weight_diff                      0.449685    1.567818
1  SPEED_LIMIT_striker                      0.044645    1.045656

Can you translate the speed limit odds ratio into plain English?

### Feature engineering: Speed limits

Honestly, that's a pretty bad way to go about things. What's more fun is if we **translate speed limits into bins.**

First, we'll use `pd.cut` to assign each speed limit a category.

In [66]:
speed_bins = [-np.inf, 10, 20, 30, 40, 50, np.inf]
merged['speed_bin'] = pd.cut(merged.SPEED_LIMIT_struck, bins=speed_bins)
merged[['SPEED_LIMIT_striker', 'speed_bin']].head(10)

SPEED_LIMIT_striker     speed_bin
0                   25  (20.0, 30.0]
1                   25  (20.0, 30.0]
2                   25  (20.0, 30.0]
3                   25  (20.0, 30.0]
4                   25  (20.0, 30.0]
5                    5  (-inf, 10.0]
6                   45  (40.0, 50.0]
7                   45  (40.0, 50.0]
8                   45  (40.0, 50.0]
9                   45  (40.0, 50.0]

Then we'll one-hot encode around 20-30mph speed limits.

In [67]:
speed_dummies = pd.get_dummies(merged.speed_bin, 
                               prefix='speed').drop('speed_(20.0, 30.0]', axis=1)
speed_dummies.head()

speed_(-inf, 10.0]  speed_(10.0, 20.0]  speed_(30.0, 40.0]  \
0                   0                   0                   0   
1                   0                   0                   0   
2                   0                   0                   0   
3                   0                   0                   0   
4                   0                   0                   0   

   speed_(40.0, 50.0]  speed_(50.0, inf]  
0                   0                  0  
1                   0                  0  
2                   0                  0  
3                   0                  0  
4                   0                  0

## Running a regression

I like this layout for creating `train_df`, it allows us to easily add dummies and do a little replacing/encoding when we're building binary features like for sex.

> If the below gives you an error, it's because `SEX_CODE` is already a number. In that case, just remove `.replace({'M': 1, 'F': 0, 'U': np.nan })`.

In [68]:
# Start with our normal features
train_df = pd.DataFrame({
    'weight_diff': merged.weight_diff,
    'sex': merged.SEX_CODE,#.replace({'M': 1, 'F': 0, 'U': np.nan }),
    'had_fatality': merged.had_fatality,
})
# Add the one-hot encoded features
train_df = train_df.join(speed_dummies)
train_df = train_df.join(surf_dummies)
# Drop missing values
train_df = train_df.dropna()
train_df.head()

weight_diff sex  had_fatality  speed_(-inf, 10.0]  speed_(10.0, 20.0]  \
0        0.124   F             0                   0                   0   
1       -0.124   F             0                   0                   0   
2        0.100   F             0                   0                   0   
3       -0.100   F             0                   0                   0   
4       -0.100   M             0                   0                   0   

   speed_(30.0, 40.0]  speed_(40.0, 50.0]  speed_(50.0, inf]  surface_Ice  \
0                   0                   0                  0            0   
1                   0                   0                  0            0   
2                   0                   0                  0            0   
3                   0                   0                  0            0   
4                   0                   0                  0            0   

   surface_Mud, Dirt, Gravel  surface_Not Applicable  surface_Other  \
0                          0                       0              0   
1                          0                       0              0   
2                          0                       0              0   
3                          0                       0              0   
4                          0                       0              0   

   surface_Snow  surface_Unknown  surface_Wet  
0             0                0            0  
1             0                0            0  
2             0                0            0  
3             0                0            0  
4             0                0            0

Describe the impact of the different variables in simple language. What has the largest impact?

## Now you pick the features

Up above you have examples of:

* Creating features from numbers (speed limits)
* Creating features from 0/1 (sex)
* Creating features from binning numbers that are one-hot encoded (speed limit bins - `speed_bins`)
* Creating features from categories that are one-hot encoded (surface - `surf_dummies`

What else do you think matters? Try to plug in more features and see if you can get anything interesting.

> * **Hot tip:** The funniest/most interesting thing feature you can add is also the dumbest. Ask me about it in #algorithms if you end up getting down here.